smart-contracts/contracts/cl-vault/src/rewards/distribution.rs

```diff
@@ -159,101 +81,188 @@ pub fn handle_collect_spread_rewards_reply(
 ) -> Result<Response, ContractError> {
     // after we have collected both the spread rewards and the incentives, we can distribute them over the share holders
     // we don't need to save the rewards here again, just pass it to update rewards
-    let data: Result<MsgCollectSpreadRewardsResponse, ContractError> = data
-        .into_result()
-        .map_err(StdError::generic_err)?
-        .data
-        .map(|b| Ok(b.try_into()?))
-        .unwrap_or(Ok(MsgCollectSpreadRewardsResponse {
-            collected_spread_rewards: vec![],
-        }));
-
-    let response: MsgCollectSpreadRewardsResponse = data?;
-    let mut rewards = CURRENT_REWARDS.load(deps.storage)?;
-    rewards.update_rewards(&response.collected_spread_rewards)?;
+    let data: MsgCollectSpreadRewardsResponse = data.try_into()?;
+
+    let mut response_coin_list = CoinList::new();
+    response_coin_list.merge_osmocoins(data.collected_spread_rewards)?;
 
-    // calculate and update the strategist fee
+    // calculate the strategist fee and remove the share at source
     let vault_config = VAULT_CONFIG.load(deps.storage)?;
-    let strategist_fee = rewards.sub_ratio(vault_config.performance_fee)?;
+    let strategist_fee = response_coin_list.sub_ratio(vault_config.performance_fee)?;
     STRATEGIST_REWARDS.update(deps.storage, |old| old.add(strategist_fee))?;
 
+    let mut rewards = CURRENT_REWARDS.load(deps.storage)?;
+    rewards.update_rewards_coin_list(response_coin_list)?;
+
     CURRENT_REWARDS.save(deps.storage, &rewards)?;
 
-    Ok(Response::new().add_attribute(
-        "collected_spread_rewards",
-        format!("{:?}", response.collected_spread_rewards),
-    ))
+    // TODO add a nice response
+    Ok(Response::new()
+        .add_attribute("method", "reply")
+        .add_attribute("action", "handle_collect_spread_rewards"))
 }
 
-pub fn execute_distribute_rewards(
+pub fn execute_auto_compound_swap(
     deps: DepsMut,
-    _env: Env,
-    amount_of_users: Uint128,
+    env: Env,
+    force_swap_route: bool,
+    mut swap_routes: Vec<AutoCompoundAsset>,
 ) -> Result<Response, ContractError> {
-    let mut rewards_status = REWARDS_STATUS.load(deps.storage).unwrap();
-    if rewards_status != RewardsStatus::Distributing {
-        return Err(ContractError::IsNotDistributing {});
+    // auto compound admin
+    let dex_router = DEX_ROUTER.may_load(deps.storage)?;
+    if swap_routes.is_empty() {
```

Comment:
Are we sure we want to error if any route is provided? Should the cw-dex-router being able to provide one by default? At this point from a design standpoint is like this entrypoint is a "swap_external_to_pool_denom", which prepares the vault for autocompounding idle assets and rewards, but not directly related to the autocompound action.